File: 15. Shifting Letters II.java
class Solution {
    public String shiftingLetters(String s, int[][] shifts) {
        int n = s.length();
        int[] shiftArray = new int[n + 1]; // Initialize a shift array with one extra space for prefix sums

        // Process the shifts array to calculate net shifts
        for (int[] shift : shifts) {
            int start = shift[0];
            int end = shift[1];
            int direction = shift[2];
            int delta = direction == 1 ? 1 : -1;
            shiftArray[start] += delta;
            shiftArray[end + 1] -= delta;
        }

        // Compute the prefix sum to get the net shift for each index
        for (int i = 1; i < n; i++) {
            shiftArray[i] += shiftArray[i - 1];
        }

        // Build the final shifted string
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < n; i++) {
            // Calculate the effective shift (mod 26 for wrapping around)
            int effectiveShift = shiftArray[i] % 26;
            char newChar = (char) ((s.charAt(i) - 'a' + effectiveShift + 26) % 26 + 'a'); // Add 26 to handle negative shifts
            result.append(newChar);
        }

        return result.toString();
    }

    // Example usage
    public static void main(String[] args) {
        Solution solution = new Solution();
        String s = "abc";
        int[][] shifts = {{0, 1, 0}, {1, 2, 1}};
        System.out.println(solution.shiftingLetters(s, shifts)); // Output: "zac"
    }
}

File: 15. Shifting Letters II.py
class Solution(object):
    def shiftingLetters(self, s, shifts):
        """
        :type s: str
        :type shifts: List[List[int]]
        :rtype: str
        """
        n = len(s)
        shift_array = [0] * (n + 1)  # Initialize a shift array with one extra space for prefix sums

        # Process the shifts array to calculate net shifts
        for start, end, direction in shifts:
            delta = 1 if direction == 1 else -1
            shift_array[start] += delta
            shift_array[end + 1] -= delta

        # Compute the prefix sum to get the net shift for each index
        for i in range(1, n):
            shift_array[i] += shift_array[i - 1]

        # Build the final shifted string
        result = []
        for i in range(n):
            # Calculate the effective shift (mod 26 for wrapping around)
            effective_shift = shift_array[i] % 26
            new_char = chr((ord(s[i]) - ord('a') + effective_shift) % 26 + ord('a'))
            result.append(new_char)

        return ''.join(result)
